Userland/Applications/PixelPaint/Tool.h

```diff
@@ -19,9 +19,37 @@ class Tool {
 public:
     virtual ~Tool();
 
-    virtual void on_mousedown(Layer&, GUI::MouseEvent&, GUI::MouseEvent&) { }
-    virtual void on_mousemove(Layer&, GUI::MouseEvent&, GUI::MouseEvent&) { }
-    virtual void on_mouseup(Layer&, GUI::MouseEvent&, GUI::MouseEvent&) { }
+    struct MouseEvent {
+        enum class Action {
+            MouseDown,
+            MouseMove,
+            MouseUp
+        };
+
+        MouseEvent(Action action, GUI::MouseEvent& layer_event, GUI::MouseEvent& image_event, GUI::MouseEvent& raw_event)
+            : m_action(action)
+            , m_layer_event(layer_event)
+            , m_image_event(image_event)
+            , m_raw_event(raw_event)
+        {
+        }
+
+        Action action() const { return m_action; }
+        GUI::MouseEvent const& layer_event() const { return m_layer_event; }
+        GUI::MouseEvent const& image_event() const { return m_image_event; }
+        GUI::MouseEvent const& raw_event() const { return m_raw_event; }
+
+    private:
```

Comment:
structs don't have public/private, I think you meant class

structs allow unrestricted access to all members and usually have no getters/setters

Reply:
That's not true, structs have public and private access just like classes, only differences is that they're just public by default

Reply:
anyway, use a class if you gonna use private members....